src/go/wsl-helper/pkg/wsl-utils/version.go

```diff
@@ -0,0 +1,290 @@
+	"golang.org/x/sys/windows"
+)
+
+// WSLInfo describes the current (online) WSL installation.
+type WSLInfo struct {
+	Installed bool           `json:"installed"`  // Whether WSL is considered to be installed.
+	Inbox     bool           `json:"inbox"`      // Whether WSL was shipped in-box or from the MS Store/MSIX
+	HasKernel bool           `json:"has_kernel"` // Whether WSL has a kernel installed
+	Version   PackageVersion `json:"version"`    // Installed WSL version (only for store version)
+}
+
+const (
+	kPackageFamily            = "MicrosoftCorporationII.WindowsSubsystemForLinux_8wekyb3d8bbwe" // spellcheck-ignore-line
+	kMsiUpgradeCode           = "{1C3DB5B6-65A5-4EBC-A5B9-2F2D6F665F48}"
+	PACKAGE_INFORMATION_BASIC = 0x00000000
+	PACKAGE_INFORMATION_FULL  = 0x00000100
+)
+
+var (
+	dllKernel32                = windows.NewLazySystemDLL("kernel32.dll")
+	getPackagesByPackageFamily = dllKernel32.NewProc("GetPackagesByPackageFamily")
+	openPackageInfoByFullName  = dllKernel32.NewProc("OpenPackageInfoByFullName")
+	closePackageInfo           = dllKernel32.NewProc("ClosePackageInfo")
+	getPackageInfo             = dllKernel32.NewProc("GetPackageInfo")
+
+	dllMsi                 = windows.NewLazySystemDLL("msi.dll")
+	msiEnumRelatedProducts = dllMsi.NewProc("MsiEnumRelatedProductsW")
+
+	// kWSLExeOverride is a context key to override how we run wsl.exe for
+	// testing.
+	kWSLExeOverride = &struct{}{}
+	// kUpgradeCodeOverride is a context key to override the MSI file to look for.
+	kUpgradeCodeOverride = &struct{}{}
+)
+
+// getPackageNames returns the package names for the given package family.
+func getPackageNames(packageFamily string) ([]string, error) {
+	var count, bufferLength uint32
+	packageFamilyBytes, err := windows.UTF16PtrFromString(packageFamily)
+	if err != nil {
+		return nil, fmt.Errorf("error allocating package family name: %w", err)
+	}
+	rv, _, err := getPackagesByPackageFamily.Call(
+		uintptr(unsafe.Pointer(packageFamilyBytes)),
+		uintptr(unsafe.Pointer(&count)),
+		uintptr(unsafe.Pointer(nil)),
+		uintptr(unsafe.Pointer(&bufferLength)),
+		uintptr(unsafe.Pointer(nil)),
+	)
+	switch rv {
+	case uintptr(windows.ERROR_SUCCESS):
+		break
+	case uintptr(windows.ERROR_INSUFFICIENT_BUFFER):
+		// This is expected: we didn't provide any buffer
+		break
+	default:
+		return nil, fmt.Errorf("error getting buffer size: %w: %w", windows.Errno(rv), err)
+	}
+
+	packageNames := make([]uintptr, count)
+	packageNameBuffer := make([]uint16, bufferLength)
+
+	rv, _, err = getPackagesByPackageFamily.Call(
+		uintptr(unsafe.Pointer(packageFamilyBytes)),
+		uintptr(unsafe.Pointer(&count)),
+		uintptr(unsafe.Pointer(unsafe.SliceData(packageNames))),
+		uintptr(unsafe.Pointer(&bufferLength)),
+		uintptr(unsafe.Pointer(unsafe.SliceData(packageNameBuffer))),
+	)
+	if rv != uintptr(windows.ERROR_SUCCESS) {
+		return nil, fmt.Errorf("error getting package names: %w: %w", windows.Errno(rv), err)
+	}
+
+	result := make([]string, 0, count)
```

Comment:
So I typically avoid using `append` when I know the final size of a slice in advance, and typically do:

```
result := make([]string, count, count)
...
for i, ptr := range slice {
  result[i] = windows.UTF16PtrToString(ptr)
}
```

It's likely that this is what `append` does under the hood for arrays < some size N, but I prefer to avoid an `append` operation if it isn't needed.